client.py

```diff
@@ -161,97 +200,92 @@ def close(self):
             self._evfail.clear()
             _reader = self._reader()
             try:
-                serv = socket.getaddrinfo(self._server, self._port)[0][-1]  # server read
                 # If server is down OSError e.args[0] = 111 ECONNREFUSED
+                serv = socket.getaddrinfo(self._server, self._port)[0][-1]  # server read
                 self._sock.connect(serv)
-                self._sock.setblocking(False)
-                # Start reading before server can send: can't send until it
-                # gets ID.
-                loop.create_task(_reader)
-                # Server reads ID immediately, but a brief pause is probably wise.
-                await asyncio.sleep_ms(50)
-                # No need for lock yet.
-                await self._send(self._my_id)  # Can throw OSError
-            except OSError:
-                if init:
-                    await self.bad_server()
-            else:
-                # Improved cancellation code contributed by Kevin Köck
-                # Note _writer pauses before 1st tx
-                _writer = self._writer()
-                loop.create_task(_writer)
+            except OSError as e:
+                if e.args[0] in (errno.ECONNABORTED, errno.ECONNRESET, errno.ECONNREFUSED):
+                    if init:
+                        await self.bad_server()
+                    continue  # temporary server outage?
+            self._sock.setblocking(False)
+            # Start reading before server can send: can't send until it
+            # gets ID.
+            loop.create_task(_reader)
+            # Server reads ID immediately, but a brief pause is probably wise.
+            await asyncio.sleep_ms(50)
+            # last_mid = self._recv_mid
+            self._ack_mid = -1
+            preheader = bytearray(5)
+            preheader[0] = 0x2C  # mid but in this case protocol identifier but will receive an ACK with mid 0x2C
+            preheader[1] = 0  # header length
+            preheader[2] = len(self._my_id) & 0xFF
+            preheader[3] = (len(self._my_id) >> 8) & 0xFF  # allows for 65535 message length
+            preheader[4] = 0xFF  # clean connection, shows if device has been reset or just a wifi outage
+            preheader = ubinascii.hexlify(preheader)
+            ok = await self._write(preheader, None, self._my_id, True, 0x2C, init=True)
```

Comment:
Sending client_id as a qos message. Ensures that the server responds correctly before sending the first message.